sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/NestedColumnAliasing.scala

```diff
@@ -0,0 +1,193 @@
+   * GetArrayItem, the nested field substitution will not be performed.
+   */
+  private def getNestedFieldReferences(plan: LogicalPlan): Seq[GetStructField] = {
+    def helper(e: Expression): Seq[GetStructField] = e match {
+      case f @ GetStructField(child, _, _) if isGetStructFieldOrAttr(child) => Seq(f)
+      case es if es.children.nonEmpty => es.children.flatMap(e => helper(e))
+      case _ => Seq.empty
+    }
+
+    def isGetStructFieldOrAttr(e: Expression): Boolean = e match {
+      case GetStructField(child, _, _) => isGetStructFieldOrAttr(child)
+      case _: AttributeReference => true
+      case _ => false
+    }
+
+    plan.expressions.flatMap(helper)
+  }
+
+  /**
+   * Return two maps in order to replace nested fields to aliases.
+   *
+   * 1. GetStructField -> Alias: A new alias is created for each nested field.
+   * 2. ExprId -> Seq[Alias]: A reference attribute has multiple aliases pointing it.
+   */
+  private def getAliasSubMap(plans: LogicalPlan*)
+    : Option[(Map[GetStructField, Alias], Map[ExprId, Seq[Alias]])] = {
+    val rootReferences = plans.map(getRootReferences).reduce(_ ++ _)
+    val nestedFieldReferences = plans.map(getNestedFieldReferences).reduce(_ ++ _)
+
+    val aliasSub = nestedFieldReferences
+      .filter(!_.references.subsetOf(rootReferences))
+      .groupBy(_.references.head)
+      .flatMap { case (attr: Attribute, nestedFields: Seq[GetStructField]) =>
+        // Each expression can contain multiple nested fields.
+        // Note that we keep the original names to deliver to parquet in a case-sensitive way.
+        val nestedFieldToAlias = nestedFields.distinct.map { f =>
+          val exprId = NamedExpression.newExprId
+          (f, Alias(f, s"_gen_alias_${exprId.id}")(exprId, Seq.empty, None))
+        }
+
+        if (nestedFieldToAlias.nonEmpty &&
+            nestedFieldToAlias.length < totalFieldNum(attr.dataType)) {
+          Some(attr.exprId -> nestedFieldToAlias)
+        } else {
+          None
+        }
+      }
+
+    if (aliasSub.isEmpty) {
+      None
+    } else {
+      Some((aliasSub.values.flatten.toMap, aliasSub.map(x => (x._1, x._2.map(_._2)))))
+    }
+  }
+
+  /**
+   * Return total number of fields of this type. This is used as a threshold to use nested column
+   * pruning. It's okay to underestimate. If the number of reference is bigger than this, the parent
+   * reference is used instead of nested field references.
+   */
+  private def totalFieldNum(dataType: DataType): Int = dataType match {
+    case _: AtomicType => 1
+    case StructType(fields) => fields.map(f => totalFieldNum(f.dataType)).sum
+    case ArrayType(elementType, _) => totalFieldNum(elementType)
+    case MapType(keyType, valueType, _) => totalFieldNum(keyType) + totalFieldNum(valueType)
+    case _ => 1 // UDT and others
+  }
+
+  /**
+   * Return all generated aliases. Although this is used in testing, this utility function can
+   * be used for other optimizations.
```

Comment:
IMHO it'd be better to move this into `NestedColumnAliasingSuite` now. Then, we move back when reusing this in other optimization?